src/engines/pallet_engine/mod.rs

```diff
@@ -0,0 +1,601 @@
+///  which will overwrite `input` with the processed `output`
+pub struct PalletEngine {
+	/// Input source to PalletEngine - This will be the path to the runtime/src/lib.rs
+	/// In code, this is never touched directly until the final step where it's overwritten
+	/// Interim, it's treated as immutable data
+	input: PathBuf,
+	/// Stores the details of the runtime pallets built from input
+	details: PalletDetails,
+	/// Stores imports necessary to make new items available
+	imports: ImportDetails,
+	/// This stores the path to the runtime file being processed
+	/// User must have read/write permissions for potentially destructive editing
+	/// All insertions are ultimately written here
+	output: PathBuf,
+	// /// This stores the path to the runtime manifest file runtime/Cargo.toml
+	// manifest: PathBuf,
+	/// State
+	state: State,
+	/// Cursor for tracking where we are in the output
+	cursor: usize,
+}
+impl Drop for PalletEngine {
+	fn drop(&mut self) {
+		let output_dir = self.output.parent().unwrap();
+		let _ = fs::remove_dir_all(output_dir);
+	}
+}
+
+/// PalletDetails is data generated after parsing of a given `input` runtime file
+/// This will make observations as to which pallets are there, if there are instances of the pallets
+/// among details such as the span of the construct_runtime! macro, the type of runtime declaration etc.
+struct PalletDetails {
+	/// Number of pallets in the runtime, changes based on pallets added to output
+	pallets: Vec<ReadPalletEntry>,
+	/// construct_runtime! macro span start location. Any pallet that's being added
+	/// should search, uptil this point to make sure existing pallets do not conflict
+	/// Note: Read-only from self.input
+	crt_start: usize,
+	/// construct_runtime! macro span end location.
+	/// For changes that happen after construct_runtime! is edited
+	/// Note: Read-only from self.input
+	crt_end: usize,
+	/// Total number of lines in input. Useful for inserting lines.
+	file_end: usize,
+	/// Type of runtime declaration being processed
+	declaration: RuntimeDeclaration,
+}
+struct ImportDetails {
+	/// On reading the source file, we obtain `last_import` which is the ending line of final import
+	/// statement, and also from where additional pallet imports must be added
+	last_import: usize,
+}
+// Public API
+impl PalletEngine {
+	/// Query the output path
+	pub fn output(&self) -> &Path {
+		&self.output
+	}
+	/// Consume self merging `output` and `input`
+	/// Call this to finalize edits
+	pub fn merge(mut self) -> anyhow::Result<()> {
+		// TODO: since we are not interacting with any post-CRT items, this is ok
+		&mut self.append_lines_from(self.details.crt_end + 1, self.details.file_end)?;
+		fs::copy(&self.output, &self.input)?;
+		fs::remove_file(&self.output);
+		Ok(())
+	}
+	/// Create a new PalletEngine
+	pub fn new(input: &PathBuf) -> anyhow::Result<Self> {
+		let tmp_dir = PathBuf::from(format!("/tmp/pallet_engine_{}", uuid::Uuid::new_v4()));
```

Comment:
Is not better to create and remove the `/tmp/pallet_engine` than having to add a uuid?

Reply:
but the tmp folder is removed after no?

Reply:
No it's not. At least not in code. We are relying on linux system to clean it up by placing it under `/tmp`